zulip_bots/zulip_bots/test_lib.py

```diff
@@ -70,29 +71,58 @@ def check_expected_responses(self, expectations, expected_method='send_reply',
                 message = {'content': m, 'type': "private", 'display_recipient': recipient,
                            'sender_email': email, 'sender_id': sender_id,
                            'sender_full_name': sender_full_name}
-                self.assert_bot_response(message=message, response=response,
-                                         expected_method=expected_method, state_handler=state_handler)
+                self.assert_bot_responses(message, (response, expected_method),
+                                          state_handler=state_handler)
             if type != "private":
                 message = {'content': m, 'type': "stream", 'display_recipient': recipient,
                            'subject': subject, 'sender_email': email, 'sender_id': sender_id,
                            'sender_full_name': sender_full_name}
-                self.assert_bot_response(message=message, response=response,
-                                         expected_method=expected_method, state_handler=state_handler)
+                self.assert_bot_responses(message, (response, expected_method),
+                                          state_handler=state_handler)
 
-    def call_request(self, message, expected_method, response, state_handler):
-        # type: (Dict[str, Any], str, Dict[str, Any], Optional[StateHandler]) -> None
+    def call_request(self, message, *responses, **kwargs):
+        # type: (Dict[str, Any], *Tuple[Dict[str, Any], str], **Any) -> None
+        state_handler = kwargs.pop('state_handler', None)
         if state_handler is None:
             state_handler = StateHandler()
-        # Send message to the concerned bot
-        self.message_handler.handle_message(message, self.MockClass(), state_handler)
 
-        # Check if the bot is sending a message via `send_message` function.
-        # Where response is a dictionary here.
+        # Mock BotHandler; get instance
+        bot_handler = self.MockClass()
         instance = self.MockClass.return_value
-        if expected_method == "send_message":
-            instance.send_message.assert_called_with(response)
-        else:
-            instance.send_reply.assert_called_with(message, response['content'])
+
+        # Send message to the bot
+        try:
+            self.message_handler.handle_message(message, bot_handler, state_handler)
+        except KeyError as key_err:
+            raise Exception("Message tested likely required key {}.".format(key_err))
```

Comment:
Did you mean to include the `.assert_called_with` in there?

This is essentially the same issue as later in the code, with a `print` outputting not in the same context.

I've been exploring this a bit, and I believe that, at least in python3 (even without a `from key_err`), raising within an except still gives a nice 'context', such as when I remove a line from testing tictactoe:
```
ERROR: test_bot (zulip_bots.bots.tictactoe.test_tictactoe.TestTictactoeBot)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/neil/repos/python-zulip-api/zulip_bots/zulip_bots/test_lib.py", line 95, in call_request
    self.message_handler.handle_message(message, bot_handler, state_handler)
  File "/home/neil/repos/python-zulip-api/zulip_bots/zulip_bots/bots/tictactoe/tictactoe.py", line 282, in handle_message
    original_sender = message['sender_email']
KeyError: 'sender_email'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/neil/repos/python-zulip-api/zulip_bots/zulip_bots/bots/tictactoe/test_tictactoe.py", line 48, in test_bot
    'send_message', state_handler=state)
  File "/home/neil/repos/python-zulip-api/zulip_bots/zulip_bots/test_lib.py", line 174, in assert_bot_response
    self.call_request(message, (response, expected_method), **kwargs)
  File "/home/neil/repos/python-zulip-api/zulip_bots/zulip_bots/test_lib.py", line 97, in call_request
    raise Exception("Message tested likely required key {}.".format(key_err))
Exception: Message tested likely required key 'sender_email'.
```
though in python2 this becomes only:
```
ERROR: test_bot (zulip_bots.bots.tictactoe.test_tictactoe.TestTictactoeBot)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "zulip_bots/bots/tictactoe/test_tictactoe.py", line 48, in test_bot
    'send_message', state_handler=state)
  File "zulip_bots/test_lib.py", line 174, in assert_bot_response
    self.call_request(message, (response, expected_method), **kwargs)
  File "zulip_bots/test_lib.py", line 97, in call_request
    raise Exception("Message tested likely required key {}.".format(key_err))
Exception: Message tested likely required key 'sender_email'.
```
So, I agree with you that context is lost in this case at least. Can we just suggest using python3? :)